x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/RuleQueryBuilder.java

```diff
@@ -245,16 +240,35 @@ public void onFailure(Exception e) {
         });
 
         QueryBuilder newOrganicQuery = organicQuery.rewrite(queryRewriteContext);
```

Comment:
We need to rewrite until the query doesn't change and we don't know how many rounds are needed. 
So we could start with something like:
``` 
QueryBuilder newOrganicQuery = organicQuery.rewrite(queryRewriteContext);
if (newOrganicQuery != organicQuery) {
  // return a new rule query builder
}
```

Reply:
Right, this is what I'm doing, but I have to return a new query builder with the supplier information so didn't do the equality check on the rewritten organic query.

Reply:
This is not enough, we need to rewrite the organic query until it doesn't change. That might require multiple rounds, see how the `BoolQueryBuilder` does.

Reply:
OK, I see what you're saying. I added a check, but at the beginning of this method before the async call is made. Once we get to this point we'll return a new RuleQueryBuilder regardless, always with a new organic query. Does this make sense?

Reply:
Since the ultimate goal of the query is to rewrite into a `PinnedQueryBuilder` we don't need to do the inner rewrite of the organic query. We can leave this rewrite to the next round (when the query becomes a  `PinnedQueryBuilder`), sorry I miss the fact that the query is only used as a transient state for the rewrite in this PR.